Pass Hostname to Certificate Verification

jens-diewald opened the pull request:
Introduce verification of host names as mentioned in 1) in #56 

I suppose using the hostname set in set_server_hostname for this is fine? The [documentation](https://wintls.dev/classes.html#_CPPv4N5boost6wintls6stream19set_server_hostnameERKNSt6stringE) already states that this is also used for certificate verification.

The necessary conversion to wstring is unfortunate, but i do not see a better way to do this..?

Ideally there would be a test case for this and possibly more cases in certificate_test.cpp, i guess.
Just taking certificates from badssl.com would be a simple approach. But they will expire at some point and would have to be renewed which depends on the continued avilability of that site.

laudrup commented:
Thanks a lot for this.

I would really appreciate it if you could add a test for this. Both for failing and successful hostname validation.

I hope it should be possible to extend the existing tests.
Thanks.

jens-diewald commented:
I have added a simple test case. I am not super happy with it, as it really only checks host name verification. I think it would be nice to have a more general facility to test certificate verification, but i am not sure how that could be done right now and it would certainly be more work.

Also i have adjusted my original commit to your indentation and brace wrapping. I did not realize this before. Having a .clang-format file in the repository would help a lot to avoid such mistakes. Could you maybe provide one, or do you happen to use some other form of formatting rules?

By the way: The Unittests in error_test.cpp that are checking error messages as strings fail for me, as the strings are in german for me.

laudrup commented on test/certificate_test.cpp:
```diff
@@ -72,3 +73,16 @@ TEST_CASE("import private key") {
 }
+
+TEST_CASE("verify certificate host name")
+{
```
Really minor thing but please move the starting brace to the previous line

laudrup commented:
> I have added a simple test case. I am not super happy with it, as it really only checks host name verification. I think it would be nice to have a more general facility to test certificate verification, but i am not sure how that could be done right now and it would certainly be more work.
> 

Thanks a lot. It's definitely better than nothing.

A more general way of testing certificate verification would definitely be a good idea but I agree that it is out of scope for this change. You are of course very welcome to try and implement it ;-)

> Also i have adjusted my original commit to your indentation and brace wrapping. I did not realize this before. Having a .clang-format file in the repository would help a lot to avoid such mistakes. Could you maybe provide one, or do you happen to use some other form of formatting rules?
> 

I actually have a clang-format I just haven't commited it for some reason. I have a bit mixed opinions about whether it's a good idea. Auto formatting code is mostly fine but really annoying in the cases where it doesn't work.

I'll try to see if I can add what I have. I probably won't do anything to enforce it's usage though.

> By the way: The Unittests in error_test.cpp that are checking error messages as strings fail for me, as the strings are in german for me.

That's pretty bad. I hadn't considered that and tests definitely shouldn't depend on the locale of the OS.

I guess I could either see if I could find a way of testing that the errors actually do end up as `SECURITY_STATUS` errors which is really the point of the tests or do some ugly changing of locale for the test process.

If you have any idea on how to solve this please do let me know. In any case it would be nice if you could create an issue for this so I don't forget about it.

Anyway, I think I can merge this if you feel confident it should be OK?

jens-diewald commented:
> I actually have a clang-format I just haven't commited it for some reason. I have a bit mixed opinions about whether it's a good idea. Auto formatting code is mostly fine but really annoying in the cases where it doesn't work.
> 
> I'll try to see if I can add what I have. I probably won't do anything to enforce it's usage though.

That would be great! I had to actively work against my auto formatting and still did miss the one brace you mentioned above ;)

> > By the way: The Unittests in error_test.cpp that are checking error messages as strings fail for me, as the strings are in german for me.
> 
> That's pretty bad. I hadn't considered that and tests definitely shouldn't depend on the locale of the OS.
> 
> I guess I could either see if I could find a way of testing that the errors actually do end up as `SECURITY_STATUS` errors which is really the point of the tests or do some ugly changing of locale for the test process.
> 
> If you have any idea on how to solve this please do let me know. In any case it would be nice if you could create an issue for this so I don't forget about it.

I will create an issue. Off the top of my head i have no good idea either...

> Anyway, I think I can merge this if you feel confident it should be OK?

Yes, should be okay :+1:

laudrup commented:
@jens-diewald 

It all looks very good. One final thing.

In order to merge this I think it should be squashed into a single commit. It is one change after all.

It seems like I had disabled squashing and merging in github but I have enabled that now. Don't know if that affects you.

Could you please squash the commits into a single commit and provide a bit more descriptive commit message?

A header like "Pass Hostname to Certificate Verification" is fine, but please add a few lines on why this change was necessary. I like descriptive commit messages.

Thanks once again.

jens-diewald commented:
I have rebased my Branch to the current master, squashed the commits and added a descriptive message.

However, i also started thinking about the host name encoding.
Do you expect wintls to support non-Ascii encodings (i.e. IDNs)?
My usage of `std::mbstowcs` converts the host name based on the current locale. [According to the documentation](https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/ns-wincrypt-httpspolicycallbackdata) it would be fine to pass non-Ascii strings to `CertVerifyCertificateChainPolicy` and they would internally be converted to punycode and handled reasonably.
For setting the SNI host name via `InitializeSecurityContextA`, i expect that always ANSI (CP1252) encoding is assumed. But i can not find any documentation if that function would actuall handle non-Ascii strings... The corresponding parameter `pszTargetName` is not even described [here](https://docs.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-initializesecuritycontexta) and there is no relevant mention for [the W variant](https://docs.microsoft.com/en-us/windows/win32/api/sspi/nf-sspi-initializesecuritycontextw).
Afert all, i am not even sure, `asio::ip::tcp::resolver` supports non-Ascii host names? It does not seem to work for me, even though it seems it seems to be implemented in terms of [getaddrinfo](https://docs.microsoft.com/en-us/windows/win32/api/ws2tcpip/nf-ws2tcpip-getaddrinfo) for me, which should support ANSI host names.
The Asio documentation does not say anything regarding encoding. Do you know more about this?

If you think only support Ascii host names should be supported in wintls, i still think this PR is good to merge.
If you expect ANSI encodings to work, i should change the `std::mbstowcs` to something else and add another test case.

jens-diewald commented:
Ah, i wrongly assumed the Windows A functions always expect CP1252. They do expect the active code page though, so using std::mbtowcs should be fine (in line with directly passing the host name to InitializeSecurityContextA).
(Oh my, why can't everything be unicode :) ).

Still, do you know if asio should be able to handle non-Ascii? It does not seem to work for me.
